File: Java/SVY21.java
import java.util.Map;
import java.util.TreeMap;

/**
 * 
 * @author cgcai
 * 
 * The SVY21 class provides functionality to convert between the SVY21 and Lat/Lon coordinate systems.
 * 
 * Internally, the class uses the equations specified in the following web page to perform the conversion.
 * http://www.linz.govt.nz/geodetic/conversion-coordinates/projection-conversions/transverse-mercator-preliminary-computations/index.aspx
 */
public class SVY21 {
	
	// Map Keys
	private static final String KEY_SVY21 = "svy21";
	private static final String KEY_LATLON = "latlon";
	private static final String KEY_EASTING = "easting";
	private static final String KEY_NORTHING = "northing";
	private static final String KEY_LONGITUDE = "longitude";
	private static final String KEY_LATITUDE = "latitude";

	private final static double radRatio = Math.PI / 180;

	/**
	 * Returns a String representing a Key in a Java Map associated with an SVY21 Northing.
	 * 
	 * @return the key associated with an SVY21 Easting in a result.
	 */
	public static String eastingKey() {
		return KEY_EASTING;
	}
	
	/**
	 * Returns a String representing a Key in a Java Map associated with Latitude.
	 * 
	 * @return the key associated with Latitude in a result.
	 */
	public static String latitudeKey() {
		return KEY_LATITUDE;
	}
	
	/**
	 * Returns a String representing a Key in a Java Map.
	 * 
	 * The presence of this key indicates that a returned Map is a Lat/Lon coordinate.
	 * 
	 * @return key associated with a Lat/Lon coordinate.
	 */
	public static String latlonCoordinateKey() {
		return KEY_LATLON;
	}
	
	/**
	 * Returns a String representing a Key in a Java Map associated with Longitude.
	 * 
	 * @return the key associated with Longitude in a result.
	 */
	public static String longitudeKey() {
		return KEY_LONGITUDE;
	}
	/**
	 * Returns a String representing a Key in a Java Map associated with an SVY21 Northing.
	 * 
	 * @return the key associated with an SVY21 Northing in a result.
	 */
	public static String northingKey() {
		return KEY_NORTHING;
	}
	/**
	 * Returns a String representing a Key in a Java Map.
	 * 
	 * The presence of this key indicates that a returned Map is an SVY21 coordinate.
	 * 
	 * @return key associated with an SVY21 coordinate.
	 */
	public static String svy21CoordinateKey() {
		return KEY_SVY21;
	}
	
	// Datum and Projection
	private final double a		= 6378137;
	private final double f		= 1 / 298.257223563;
	private final double oLat	= 1.366666;				// degrees
	private final double oLon	= 103.833333;			// degrees
	private final double No		= 38744.572;	
	private final double Eo		= 28001.642;

	private final double k		= 1;

	// Projection Constants
	private final double b;
	private final double e2, e4, e6;
	private final double A0, A2, A4, A6;

	/**
	 * Class constructor.
	 * 
	 * Computes various projection constants based on WGS84 and SVY21.
	 */
	public SVY21() {
		b = a * (1 - f);

		e2 = (2 * f) - (f * f);
		e4 = e2 * e2;
		e6 = e4 * e2;

		A0 = 1 - (e2 / 4) - (3 * e4 / 64) - (5 * e6 / 256);
		A2 = (3. / 8.) * (e2 + (e4 / 4) + (15 * e6 / 128));
		A4 = (15. / 256.) * (e4 + (3 * e6 / 4));
		A6 = 35 * e6 / 3072;
	}

	private double calcM(double lat) {
		double latR = lat * radRatio;
		double m = a * ((A0 * latR) - (A2 * Math.sin(2 * latR)) + (A4 * Math.sin(4 * latR)) - (A6 * Math.sin(6 * latR)));
		return m;
	}
	
	private double calcRho(double sin2Lat) {
		double num = a * (1 - e2);
		double denom = Math.pow(1 - e2 * sin2Lat, 3. / 2.);
		double rho = num / denom;
		return rho;
	}
	
	private double calcV(double sin2Lat) {
		double poly = 1 - e2 * sin2Lat;
		double v = a / Math.sqrt(poly);
		return v;
	}
	
	/**
	 * Computes Latitude and Longitude based on an SVY21 coordinate.
	 * 
	 * This method returns a Java Map containing exactly two entries. 
	 * Latitude can be accessed by the key returned by the static method latitudeKey().
	 * Longitude can be accessed by the key returned by the static method longitudeKey().
	 * 
	 * The returned Map contains a key defined by latlonCoordinateKey() which indicates that
	 * the contained data represents a Lat/Lon coordinate. The Map can be unpacked with the keys
	 * latitudeKey() and longitudeKey().
	 * 
	 * @param N Northing based on SVY21.
	 * @param E Easting based on SVY21.
	 * @return the conversion result as a Java Map.
	 */
	public Map<String, Double> computeLatLon(double N, double E) {
		double Nprime = N - No;
		double Mo = calcM(oLat);
		double Mprime = Mo + (Nprime / k);
		double n = (a - b) / (a + b);
		double n2 = n * n;
		double n3 = n2 * n;
		double n4 = n2 * n2;
		double G = a * (1 - n) * (1 - n2) * (1 + (9 * n2 / 4) + (225 * n4 / 64)) * radRatio;
		double sigma = (Mprime / G) * radRatio;

		double latPrimeT1 = ((3 * n / 2) - (27 * n3 / 32)) * Math.sin(2 * sigma);
		double latPrimeT2 = ((21 * n2 / 16) - (55 * n4 / 32)) * Math.sin(4 * sigma);
		double latPrimeT3 = (151 * n3 / 96) * Math.sin(6 * sigma);
		double latPrimeT4 = (1097 * n4 / 512) * Math.sin(8 * sigma);
		double latPrime = sigma + latPrimeT1 + latPrimeT2 + latPrimeT3 + latPrimeT4;

		double sinLatPrime = Math.sin(latPrime);
		double sin2LatPrime = sinLatPrime * sinLatPrime;

		double rhoPrime = calcRho(sin2LatPrime);
		double vPrime = calcV(sin2LatPrime);
		double psiPrime = vPrime / rhoPrime;
		double psiPrime2 = psiPrime * psiPrime;
		double psiPrime3 = psiPrime2 * psiPrime;
		double psiPrime4 = psiPrime3 * psiPrime;
		double tPrime = Math.tan(latPrime);
		double tPrime2 = tPrime * tPrime;
		double tPrime4 = tPrime2 * tPrime2;
		double tPrime6 = tPrime4 * tPrime2;
		double Eprime = E - Eo;
		double x = Eprime / (k * vPrime);
		double x2 = x * x;
		double x3 = x2 * x;
		double x5 = x3 * x2;
		double x7 = x5 * x2;

		// Compute Latitude
		double latFactor = tPrime / (k * rhoPrime);
		double latTerm1 = latFactor * ((Eprime * x) / 2);
		double latTerm2 = latFactor * ((Eprime * x3) / 24) * ((-4 * psiPrime2) + (9 * psiPrime) * (1 - tPrime2) + (12 * tPrime2));
		double latTerm3 = latFactor * ((Eprime * x5) / 720) * ((8 * psiPrime4) * (11 - 24 * tPrime2) - (12 * psiPrime3) * (21 - 71 * tPrime2) + (15 * psiPrime2) * (15 - 98 * tPrime2 + 15 * tPrime4) + (180 * psiPrime) * (5 * tPrime2 - 3 * tPrime4) + 360 * tPrime4);
		double latTerm4 = latFactor * ((Eprime * x7) / 40320) * (1385 - 3633 * tPrime2 + 4095 * tPrime4 + 1575 * tPrime6);
		double lat = latPrime - latTerm1 + latTerm2 - latTerm3 + latTerm4;

		// Compute Longitude
		double secLatPrime = 1. / Math.cos(lat);
		double lonTerm1 = x * secLatPrime;
		double lonTerm2 = ((x3 * secLatPrime) / 6) * (psiPrime + 2 * tPrime2);
		double lonTerm3 = ((x5 * secLatPrime) / 120) * ((-4 * psiPrime3) * (1 - 6 * tPrime2) + psiPrime2 * (9 - 68 * tPrime2) + 72 * psiPrime * tPrime2 + 24 * tPrime4);
		double lonTerm4 = ((x7 * secLatPrime) / 5040) * (61 + 662 * tPrime2 + 1320 * tPrime4 + 720 * tPrime6);
		double lon = (oLon * radRatio) + lonTerm1 - lonTerm2 + lonTerm3 - lonTerm4;

		Map<String, Double> result = new TreeMap<String, Double>();
		result.put(KEY_LATLON, 0.);
		result.put(latitudeKey(), lat / radRatio);
		result.put(longitudeKey(), lon / radRatio);
		return result;
	}
	
	/**
	 * Computes Latitude and Longitude based on an SVY21 coordinate.
	 * 
	 * This method returns a Java Map containing exactly two entries. 
	 * Latitude can be accessed by the key returned by the static method latitudeKey().
	 * Longitude can be accessed by the key returned by the static method longitudeKey().
	 * 
	 * The returned Map contains a key defined by latlonCoordinateKey() which indicates that
	 * the contained data represents a Lat/Lon coordinate. The Map can be unpacked with the keys
	 * latitudeKey() and longitudeKey().
	 * 
	 * This method is a short-hand for the functionally-identical
	 * public Map<String, Double> computeLatLon(double N, double E)
	 * 
	 * @param svy21Coordinate
	 * @return
	 */
	public Map<String, Double> computeLatLon(Map<String, Double> svy21Coordinate) {
		double northing = svy21Coordinate.get(northingKey());
		double easting = svy21Coordinate.get(eastingKey());
		return computeLatLon(northing, easting);
	}
	
	/**
	 * Computes SVY21 Northing and Easting based on a Latitude and Longitude coordinate.
	 * 
	 * This method returns a Java Map containing exactly two entries. 
	 * Northings can be accessed by the key returned by the static method northingKey().
	 * Eastings can be accessed by the key returned by the static method eastingKey().
	 * 
	 * The returned Map contains a key defined by svy21CoordinateKey() which indicates that
	 * the contained data represents an SVY21 coordinate. The Map can be unpacked with the keys
	 * northingKey() and eastingKey().
	 * 
	 * @param lat latitude in degrees.
	 * @param lon longitude in degrees.
	 * @return the conversion result as a Java Map.
	 */
	public Map<String, Double> computeSVY21(double lat, double lon) {
		double latR = lat * radRatio;
		double sinLat = Math.sin(latR);
		double sin2Lat = sinLat * sinLat;
		double cosLat = Math.cos(latR);
		double cos2Lat = cosLat * cosLat;
		double cos3Lat = cos2Lat * cosLat;
		double cos4Lat = cos3Lat * cosLat;
		double cos5Lat = cos3Lat * cos2Lat;
		double cos6Lat = cos5Lat * cosLat;
		double cos7Lat = cos5Lat * cos2Lat;

		double rho = calcRho(sin2Lat);
		double v = calcV(sin2Lat);
		double psi = v / rho;
		double t = Math.tan(latR);
		double w = (lon - oLon) * radRatio;

		double M = calcM(lat);
		double Mo = calcM(oLat);

		double w2 = w * w;
		double w4 = w2 * w2;
		double w6 = w4 * w2;
		double w8 = w6 * w2;

		double psi2 = psi * psi;
		double psi3 = psi2 * psi;
		double psi4 = psi2 * psi2;

		double t2 = t * t;
		double t4 = t2 * t2;
		double t6 = t4 * t2;

		// Compute Northing.
		double nTerm1 = w2 / 2 * v * sinLat * cosLat;
		double nTerm2 = w4 / 24 * v * sinLat * cos3Lat * (4 * psi2 + psi - t2);
		double nTerm3 = w6 / 720 * v * sinLat * cos5Lat * ((8 * psi4) * (11 - 24 * t2) - (28 * psi3) * (1 - 6 * t2) + psi2 * (1 - 32 * t2) - psi * 2 * t2 + t4);
		double nTerm4 = w8 / 40320 * v * sinLat * cos7Lat * (1385 - 3111 * t2 + 543 * t4 - t6);
		double N = No + k * (M - Mo + nTerm1 + nTerm2 + nTerm3 + nTerm4);

		// Compute Easting.
		double eTerm1 = w2 / 6 * cos2Lat * (psi - t2);
		double eTerm2 = w4 / 120 * cos4Lat * ((4 * psi3) * (1 - 6 * t2) + psi2 * (1 + 8 * t2) - psi * 2 * t2 + t4);
		double eTerm3 = w6 / 5040 * cos6Lat * (61 - 479 * t2 + 179 * t4 - t6);
		double E = Eo + k * v * w * cosLat * (1 + eTerm1 + eTerm2 + eTerm3);

		Map<String, Double> result = new TreeMap<String, Double>();
		result.put(KEY_SVY21, 0.);
		result.put(northingKey(), N);
		result.put(eastingKey(), E);
		return result;
	}
	
	/**
	 * Computes SVY21 Northing and Easting based on a Latitude and Longitude coordinate.
	 * 
	 * This method returns a Java Map containing exactly two entries. 
	 * Northings can be accessed by the key returned by the static method northingKey().
	 * Eastings can be accessed by the key returned by the static method eastingKey().
	 * 
	 * The returned Map contains a key defined by svy21CoordinateKey() which indicates that
	 * the contained data represents an SVY21 coordinate. The Map can be unpacked with the keys
	 * northingKey() and eastingKey().
	 * 
	 * This method is a short-hand for the functionally-identical
	 * public Map<String, Double> computeSVY21(double lat, double lon)
	 * 
	 * @param latlonCoordinate
	 * @return
	 */
	public Map<String, Double> computeSVY21(Map<String, Double> latlonCoordinate) {
		double latitude = latlonCoordinate.get(latitudeKey());
		double longitude = latlonCoordinate.get(longitudeKey());
		return computeSVY21(latitude, longitude);
	}
}

File: Ruby/lib/svy21.rb
# The SVY21 class provides functionality to convert between  the SVY21 and Lat/Lon coordinate systems.
#
# Internally, the class uses the equations specified in the following web page to perform the conversion.
# http://www.linz.govt.nz/geodetic/conversion-coordinates/projection-conversions/transverse-mercator-preliminary-computations/index.aspx
class SVY21
  # Ratio to convert degrees to radians.
  RAD_RATIO = Math::PI / 180.0
  
  # Datum and Projection.
  
  # Semi-major axis of reference ellipsoid.
  A = 6378137.0
  # Ellipsoidal flattening
  F = 1.0 / 298.257223563
  # Origin latitude (degrees).
  ORIGIN_LATITUDE = 1.366666
  # Origin longitude (degrees).
  ORIGIN_LONGITUDE = 103.833333
  # False Northing.
  FALSE_NORTHING = 38744.572
  # False Easting.
  FALSE_EASTING = 28001.642
  # Central meridian scale factor.
  K = 1.0
  
  # Computed Projection Constants
  
  # Naming convention: the trailing number is the power of the variable.
  
  # Semi-minor axis of reference ellipsoid.
  B = A * (1.0 - F)
  # Squared eccentricity of reference ellipsoid.
  E2 = (2.0 * F) - (F * F)
  E4 = E2 * E2
  E6 = E4 * E2
  
  # Naming convention: A0..6 are terms in an expression, not powers.
  
  A0 = 1.0 - (E2 / 4.0) - (3.0 * E4 / 64.0) - (5.0 * E6 / 256.0)
  A2 = (3.0 / 8.0) * (E2 + (E4 / 4.0) + (15.0 * E6 / 128.0))
  A4 = (15.0 / 256.0) * (E4 + (3.0 * E6 / 4.0))
  A6 = 35.0 * E6 / 3072.0
  
  # Naming convention: the trailing number is the power of the variable.
  
  N = (A - B) / (A + B)
  N2 = N * N
  N3 = N2 * N
  N4 = N2 * N2
  
  G = A * (1.0 - N) * (1.0 - N2) * (1.0 + (9.0 * N2 / 4.0) + (225.0 * N4 / 64.0)) * RAD_RATIO
  
  # M: meridian distance.
  def self.calc_m(lat)
    lat_r = lat * RAD_RATIO
    
    A * ((A0 * lat_r) - (A2 * Math.sin(2.0 * lat_r)) + (A4 * Math.sin(4.0 * lat_r)) - (A6 * Math.sin(6.0 * lat_r)))
  end
  
  # Rho: radius of curvature of meridian.
  def self.calc_rho(sin_2_lat)
    num = A * (1.0 - E2)
    denom = (1.0 - E2 * sin_2_lat) ** (3.0 / 2.0)
    
    num / denom
  end
  
  # v: radius of curvature in the prime vertical.
  def self.calc_v(sin_2_lat)
    poly = 1.0 - E2 * sin_2_lat
    
    A / Math.sqrt(poly)
  end
  
  # Computes Latitude and Longitude based on an SVY21 coordinate.
  #
  # This method returns an array object that contains two numbers,
	# latitude, accessible with [0], and
	# longitude, accessible with [1].
  #
  # @param northing Northing based on SVY21.
  # @param easting Easting based on SVY21.
  #
  # @return the conversion result as an array, [lat, lon].
  def self.svy21_to_lat_lon(northing, easting)
    n_prime = northing - FALSE_NORTHING
    m_origin = calc_m(ORIGIN_LATITUDE)
    m_prime = m_origin + (n_prime / K)
    sigma = (m_prime / G) * RAD_RATIO
    
    # Naming convention: lat_prime_t1..4 are terms in an expression, not powers.
    lat_prime_t1 = ((3.0 * N / 2.0) - (27.0 * N3 / 32.0)) * Math.sin(2.0 * sigma)
    lat_prime_t2 = ((21.0 * N2 / 16.0) - (55.0 * N4 / 32.0)) * Math.sin(4.0 * sigma)
    lat_prime_t3 = (151.0 * N3 / 96.0) * Math.sin(6.0 * sigma)
    lat_prime_t4 = (1097.0 * N4 / 512.0) * Math.sin(8.0 * sigma)
    lat_prime = sigma + lat_prime_t1 + lat_prime_t2 + lat_prime_t3 + lat_prime_t4
    
    # Naming convention: sin_2_lat_prime = "square of sin(lat_prime)" = sin(lat_prime) ** 2.0
    sin_lat_prime = Math.sin(lat_prime)
    sin_2_lat_prime = sin_lat_prime * sin_lat_prime
    
    # Naming convention: the trailing number is the power of the variable.
    rho_prime = calc_rho(sin_2_lat_prime)
    v_prime = calc_v(sin_2_lat_prime)
    psi_prime = v_prime / rho_prime
    psi_prime_2 = psi_prime * psi_prime
    psi_prime_3 = psi_prime_2 * psi_prime
    psi_prime_4 = psi_prime_3 * psi_prime
    t_prime = Math.tan(lat_prime)
    t_prime_2 = t_prime * t_prime
    t_prime_4 = t_prime_2 * t_prime_2
    t_prime_6 = t_prime_4 * t_prime_2
    e_prime = easting - FALSE_EASTING
    x = e_prime / (K * v_prime)
    x2 = x * x
    x3 = x2 * x
    x5 = x3 * x2
    x7 = x5 * x2
    
    # Compute Latitude
    # Naming convention: lat_term_1..4 are terms in an expression, not powers.
    lat_factor = t_prime / (K * rho_prime)
    lat_term_1 = lat_factor * ((e_prime * x) / 2.0)
    lat_term_2 = lat_factor * ((e_prime * x3) / 24.0) * ((-4.0 * psi_prime_2 + (9.0 * psi_prime) * (1.0 - t_prime_2) + (12.0 * t_prime_2)))
    lat_term_3 = lat_factor * ((e_prime * x5) / 720.0) * ((8.0 * psi_prime_4) * (11.0 - 24.0 * t_prime_2) - (12.0 * psi_prime_3) * (21.0 - 71.0 * t_prime_2) + (15.0 * psi_prime_2) * (15.0 - 98.0 * t_prime_2 + 15.0 * t_prime_4) + (180.0 * psi_prime) * (5.0 * t_prime_2 - 3.0 * t_prime_4) + 360.0 * t_prime_4)
    lat_term_4 = lat_factor * ((e_prime * x7) / 40320.0) * (1385.0 - 3633.0 * t_prime_2 + 4095.0 * t_prime_4 + 1575.0 * t_prime_6)
    lat = lat_prime - lat_term_1 + lat_term_2 - lat_term_3 + lat_term_4
    
    # Compute Longitude
    # Naming convention: lon_term_1..4 are terms in an expression, not powers.
		sec_lat_prime = 1.0 / Math.cos(lat)
		lon_term_1 = x * sec_lat_prime
		lon_term_2 = ((x3 * sec_lat_prime) / 6.0) * (psi_prime + 2.0 * t_prime_2)
		lon_term_3 = ((x5 * sec_lat_prime) / 120.0) * ((-4.0 * psi_prime_3) * (1.0 - 6.0 * t_prime_2) + psi_prime_2 * (9.0 - 68.0 * t_prime_2) + 72.0 * psi_prime * t_prime_2 + 24.0 * t_prime_4)
		lon_term_4 = ((x7 * sec_lat_prime) / 5040.0) * (61.0 + 662.0 * t_prime_2 + 1320.0 * t_prime_4 + 720.0 * t_prime_6)
		lon = (ORIGIN_LONGITUDE * RAD_RATIO) + lon_term_1 - lon_term_2 + lon_term_3 - lon_term_4
    
    [lat / RAD_RATIO, lon / RAD_RATIO]
  end
  
  # Computes SVY21 Northing and Easting based on a Latitude and Longitude coordinate.
  #
  # This method returns an array object that contains two numbers,
  # northing, accessible with [0], and
  # easting, accessible with [1].
  #
  # @param latitude latitude in degrees.
  # @param longitude longitude in degrees.
  #
  # @return the conversion result as an array, [northing, easting].
  def self.lat_lon_to_svy21(latitude, longitude)
    # Naming convention: sin_2_lat = "square of sin(lat)" = sin(lat) ** 2.0
    lat_r = latitude * RAD_RATIO
    sin_lat = Math.sin(lat_r)
    sin_2_lat = sin_lat * sin_lat
    cos_lat = Math.cos(lat_r)
    cos_2_lat = cos_lat * cos_lat
    cos_3_lat = cos_2_lat * cos_lat
    cos_4_lat = cos_3_lat * cos_lat
    cos_5_lat = cos_3_lat * cos_2_lat
    cos_6_lat = cos_5_lat * cos_lat
    cos_7_lat = cos_5_lat * cos_2_lat
    
    rho = calc_rho(sin_2_lat)
    v = calc_v(sin_2_lat)
    psi = v / rho
    t = Math.tan(lat_r)
    w = (longitude - ORIGIN_LONGITUDE) * RAD_RATIO
    m = calc_m(latitude)
    origin_m = calc_m(ORIGIN_LATITUDE)
    
    # Naming convention: the trailing number is the power of the variable.
    w2 = w * w
    w4 = w2 * w2
    w6 = w4 * w2
    w8 = w6 * w2
    psi2 = psi * psi
    psi3 = psi2 * psi
    psi4 = psi2 * psi2
    t2 = t * t
    t4 = t2 * t2
    t6 = t4 * t2
    
    # Compute Northing.
    # Naming convention: n_term_1..4 are terms in an expression, not powers.
    n_term_1 = w2 / 2.0 * v * sin_lat * cos_lat
    n_term_2 = w4 / 24.0 * v * sin_lat * cos_3_lat * (4.0 * psi2 + psi - t2)
    n_term_3 = w6 / 720.0 * v * sin_lat * cos_5_lat * ((8.0 * psi4) * (11.0 - 24.0 * t2) - (28.0 * psi3) * (1.0 - 6.0 * t2) + psi2 * (1.0 - 32.0 * t2) - psi * 2.0 * t2 + t4)
    n_term_4 = w8 / 40320.0 * v * sin_lat * cos_7_lat * (1385.0 - 3111.0 * t2 + 543.0 * t4 - t6)
    northing = FALSE_NORTHING + K * (m - origin_m + n_term_1 + n_term_2 + n_term_3 + n_term_4)
    
    # Compute Easting.
    # Naming convention: e_term_1..3 are terms in an expression, not powers.
    e_term_1 = w2 / 6.0 * cos_2_lat * (psi - t2)
    e_term_2 = w4 / 120.0 * cos_4_lat * ((4.0 * psi3) * (1.0 - 6.0 * t2) + psi2 * (1.0 + 8.0 * t2) - psi * 2.0 * t2 + t4)
    e_term_3 = w6 / 5040.0 * cos_6_lat * (61.0 - 479.0 * t2 + 179.0 * t4 - t6)
    easting = FALSE_EASTING + K * v * w * cos_lat * (1 + e_term_1 + e_term_2 + e_term_3)
    
    [northing, easting]
  end
end